src/MongoDB.Driver/Encryption/AutoEncryptionLibMongoController.cs

```diff
@@ -223,5 +227,39 @@ private void RestoreDbNodeInResponse(BsonDocument request, BsonDocument response
                 }
             }
         }
+
+        private void WaitForMongocryptdReady()
+        {
+            using (CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromMinutes(1)))
+            {
+                while (!cancellationTokenSource.IsCancellationRequested)
+                {
+                    var clusterDescription = _mongocryptdClient.Cluster?.Description;
+                    var mongocryptdServer = clusterDescription?.Servers?.FirstOrDefault();
+                    if (mongocryptdServer != null && mongocryptdServer.Type != ServerType.Unknown)
+                    {
+                        return;
+                    }
+                    Thread.Sleep(TimeSpan.FromMilliseconds(5));
```

Comment:
Perhaps there's a better way than sleeping, which always seems to me like a bad smell. What if, instead, we register for a `MongoDB.Driver.Core.Events.ClusterDescriptionChangedEvent` and then wait here on a signal of some sort from the event listener?  (In Java I would use a CountDownLatch(1), for example).

Reply:
I don't think you can register a handler for an event without creating a whole new client. Event handlers are tied to the client and last for the lifetime of the client.

Reply:
p.s. Unless you mean a C# event instead of a CMAP event. That might work.

We could also use Spin.Wait which has its own heuristics for how to handle waiting.

Reply:
Keep in mind that this is a temporary band-aid for dealing with SDAM delays in discovering the new mongocryptd server, so it may not be important to get it perfect (and Sleep is OK).

Perhaps we should add a comment indicating that this is a temporary workaround.

Reply:
We're in control of how the client is created (i.e., the application doesn't provide the client to us, we can create it as we wish) so we can register events.  

As for the band aid: there are no SPEC tickets scheduled for GA to fix this, so we have to do the best we can.

Reply:
The problem with registering for a CMAP event is that handlers are tied to the lifetime of the client. We only care about waiting for SDAM to discover the mongocryptd for this one block of code. There is no way to unregister an event handler.

We also discussed just checking whether opening a socket to mongocryptd succeeds, which can be done in both sync/async code without using Sleep, but I believe Dmitry tried that and still got server selection errors if we return after mongocryptd is accepting connections but before SDAM has discovered it.

I don't know that a SPEC ticket is required. It might be an implementation issue with the C# driver.

Reply:
I'm not particularly worried about the CMAP event continuing to be registered.  There shouldn't be many events delivered (one every 10 seconds? or perhaps none, since the topology doesn't change)

Reply:
Let's get resolution on this comment thread.  Sounds like everyone else so far is happy enough with a `Thread.sleep()`?

Reply:
Probably it's ok for sync code.
Done